### ISSUE TITLE
[RFC] Add `orderPredicate` to promotion rule

### ISSUE THREAD

**Issue by IKarbowiak** [saleor/saleor#12889]
## Problem

Allow defining the predicated on `Checkout` and `Order` for the promotion rule. The `orderPredicate` defines the conditions that must be fulfilled on checkout or in draft order for the promotion to apply.

Blocked by: https://github.com/saleor/saleor/issues/12539.

---

## 1. General Assumptions

- Add the `orderPredicate` predicate on promotion rule
- Allow defining the reward type - add a `rewardType` to `PromotionRule`
- Allow for the shipping discount and subtotal price discount rewards
- Allow defining the `rewardType` only for `orderPredicate`
- Ensure the discount is applied when calculating prices for checkout and order
- Only one promotion with best discount applies on one checkout/order. The rules discounts are combined within single promotion. 

### Examples
- Free shipping when you spend $50
```json
{
 "promotionRule": {
  "orderPredicate": {
   "discountedObjectPredicate": {
    "subtotalPrice": {
    	"range": {
    	  "gte": 50
    	}
    }
   }
  },
  "rewardType": "SHIPPING_DISCOUNT"
  "rewardValueType": "PERCENTAGE",
  "rewardValue": 100
 }
}
```

- 20% discount on subtotal when you spend at least $100
```json
{
 "promotionRule": {
  "orderPredicate": {
   "discountedObjectPredicate": {
    "subtotalPrice": {
    	"range": {
    	  "gte": 100
    	}
    }
   }
  },
  "rewardType": "SUBTOTAL_DISCOUNT",
  "rewardValueType": "PERCENTAGE",
  "rewardValue": 20
 }
}
```

### Acceptance Criteria

- [ ] The proper promotions applied on both checkout and draft orders

### Mutations

- Add `orderPredicate` and `rewardType` to `PromotionCreateInput`
- Add `orderPredicate` and `rewardType` to `PromotionRuleCreateInput` and `PromotionRuleUpdateInput`

- Require providing `rewardType` when `orderPredicate` is provided
- Raise an error when `rewardType` is provided for `catloguePredicate`
- Allow providing only `orderPredicate` or `catloguePredicate`. Raise an error when both value are provided.
- Raise an error when the price based predicate (`subtotalPrice`, `totalPrice`) is provided for rule with multiple channels with different currencies

```graphql
input PromotionRuleInput @doc(category: "Discounts") {
  ...
  orderPredicate: OrderPredicateInput
  rewardType: RewardTypeEnum
}


input PromotionRuleCreateInput @doc(category: "Discounts") {
  ...
  orderPredicate: OrderPredicateInput
  rewardType: RewardTypeEnum
}

input PromotionRuleUpdateInput @doc(category: "Discounts") {
  ...
  orderPredicate: OrderPredicateInput
  rewardType: RewardTypeEnum
}


input OrderPredicateInput @doc(category: "Discounts") {
  discountedObjectPredicate: DiscountedObjectPredicateInput
  discountedLineObjectPredicate: DiscountedLineObjectPredicateInput
  AND: [OrderPredicateInput!]
  OR: [OrderPredicateInput!]
}


input DiscountedObjectPredicateInput @doc(category: "Discounts"){
  country: StringFilterInput
  totalQuantity: DecimalFilterInput
  subtotalPrice: DecimalFilterInput
  totalPrice: DecimalFilterInput
  createdAt: DateTimeFilterInput
  userPredicate: UserPredicateInput
}

input UserPredicateInput @doc(category: "Users"){
  ids: [ID!]
  isStaff: Boolean
  isActive: Boolean
  email: StringFilterInput
  firstName: StringFilterInput
  lastName: StringFilterInput
  dateJoined: DateTimeFilterInput
}

input DiscountedLineObjectPredicateInput @doc(category: "Discounts"){
  quantity: IntFilterInput
  totalPrice: DecimalFilterInput
  unitPrice: DecimalFilterInput
  cataloguePredicate: CataloguePredicateInput
}

```

### Types
- Add `orderPredicate` to `PromotionRule` type
- Add `rewardType` to `PromotionRule` type

```graphql
type PromotionRule implements Node @doc(category: "Discounts"){
  ...
  orderPredicate: OrderPredicate
  rewardType: RewardTypeEnum
}


type OrderPredicate @doc(category: "Discounts"){
  discountedObjectPredicate: DiscountedObjectPredicate
  discountedLineObjectPredicate: DiscountedLineObjectPredicate
  AND: [DiscountedObjectPredicate]
  OR: [DiscountedObjectPredicate]   
}

type DiscountedObjectPredicate @doc(category: "Discounts"){
  country: StringFilter
  totalQuantity: DecimalFilter
  subtotalPrice: DecimalFilter
  totalPrice: DecimalFilter
  createdAt: DateTimeFilter
  userPredicate: UserPredicate
}

type UserPredicate @doc(category: "Users"){
  ids: [ID!]
  isStaff: Boolean
  isActive: Boolean
  email: StringFilter
  firstName: StringFilter
  lastName: StringFilter
  dateJoined: DateTimeFilter
}

type DiscountedLineObjectPredicate @doc(category: "Discounts"){
  quantity: IntFilter
  totalPrice: DecimalFilter
  unitPrice: DecimalFilter
  cataloguePredicate: CatalogueObjectsPredicate
}

enum RewardTypeEnum @doc(category: "Discounts") {
  SHIPPING_DISCOUNT
  SUBTOTAL_DISCOUNT
}
```

## 3. Database changes
- Extend `PromotionRule` with `reward_type` and `checkout_and_order_perdicate`
```python
class PromotionRule:
    ...
    checkout_and_order_predicate: JSONField
    reward_type: CharField
```

- Add `CheckoutDiscount` (similar to `OrderDiscount`) that will keep discount applied on the whole `Checkout` instead of each line

## 4. Calculations
- Create `where` filter inputs for `DiscountedObjects` for `Checkout` and `Order` that will include all fields from `DiscountedObjectPredicate`
- Create `where` filter inputs for `DiscountedObjects` for `CheckoutLine` and `OrderLine` that will include all fields from `DiscountedObjectLinePredicate`
- Add a helper function similar to `get_variants_for_predicate` that will return checkouts or orders that fulfil predicate conditions. The function should gets the checkouts/orders queryset, to operate only on interesting qs. Depends if the qs is `Checkout` or `Order` type, the function should use order or checkout filter inputs for calculations.
- Add a helper function for fetching applicable promotion rules for given `Checkout` or `Order` (similar to `fetch_active_promotion_rules`). The method should take into consideration only promotions that has defined `catalogueAndOrderPredicate`. 

### Checkout calculations
- Update checkout calculations - `_fetch_checkout_prices_if_expired` should check if any rules applies on this checkout. If multiple rules applies, we should calculate the best discount for user (rules discounts are combined within single promotion). The discount should be saved in `CheckoutDiscount` or `CheckoutLineDiscount` (if the reward type is `SPECIFIC_PRODUCT_DISCOUNT`) objects
- Add tests for mutations like `CheckoutCreate`, `CheckoutLinesAdd`, `CheckoutLineDelete`, check the conditions if prices are properly recalculated if the checkout/order does not fit the rules conditions anymore 

### Order calculations
- Update order calculations - `_fetch_order_prices_if_expired` should check if any rules applies on this order. If multiple rules applies, we should calculate the best discount for user (rules discounts are combined within single promotion). The discount should be saved in `OrderDiscount` or `OrderLineDiscount` (if the reward type is `SPECIFIC_PRODUCT_DISCOUNT`) objects.
- Add tests for mutations like `DraftOrderCreate`, `OrderLinesAdd`, `OrderLineDelete`, check the conditions if prices are properly recalculated if the checkout/order does not fit the rules conditions anymore 

- The discounts should be included in the tax calculations. It should be handled in the same way as currently. So when the discount applies to subtotal or entire object, we're sending it as a separate discount object to TaxApp. The tax app responsibility was propagate the discount on the lines. 

### Tasks
- [ ] Add `orderPredicate`
- [ ] Update promotion mutations to allow defining the `orderPredicate`
- [ ] Update promotion  rule mutations to allow defining the `orderPredicate`
- [ ] Update the order and checkout calculations to include the promotion discount

## 6. To Test

- [ ] Check with override prices
- [ ] Check if the discount is removed when the order/checkout does not fit the rule conditions anymore, e.g. items were deleted

```[tasklist]
### Tasks
- [ ] https://github.com/saleor/saleor/issues/13330
- [ ] https://github.com/saleor/saleor/issues/13331
```

**Comment by korycins** [comment]
Ref - `Checkout calculations` and `Order calculations`
(Might be obvious but I will raise it just to confirm) When we decide which promotion is the best for the user, we should also trigger the logic to re-calculate the taxes.

Not sure if this is applicable here, but when we have a voucher that applies to subtotal or entire order, we're sending it as a separate discount object to TaxApp. The tax app responsibility was propagate the discount on the lines. The reason for this was: when we have 20% voucher, for entire order, we don't know exactly how we should propagate it and we pass it to the tax app. Ref https://docs.saleor.io/docs/3.x/developer/extending/webhooks/synchronous-events/tax#checkout-taxes

**Comment by korycins** [comment]
Are we sure that we want to have a single flag for `checkoutAndOrder`? What if I would have a different approach for checkout and order here? (don't know if valid case but if we will receive such a request then it can be problematic to apply). 

If a single flag is valid and expected, maybe we should consider a better name here. Personally not a fan of having `And` in such names. In Taxes for example we have a name `sourceObject`, and the source object can be an order or checkout. 
Maybe in this case we would be able to find a single, generic word that would match checkout and order.

**Comment by IKarbowiak** [comment]
@korycins thanks for comments!
Those functions for checkout and order recalculations already calls the taxes recalculations.
Regarding the tax calculations, I think for the entire order/checkout discounts it should works like in case of vouchers. I'll add the comment about that to the RFC.

For both of models logic will be the same so we want to keep it together. We were thinking about the names about the first phase of RFC, and that was the best what comes to our minds, that will be not too long, but meaningful at the same time. Maybe we should think more about the name of this predicate 🤔.

**Comment by IKarbowiak** [comment]
Partially implemented. Rest is postponed for now.
